### PR TITLE
Groundwork for saves

### PR DESCRIPTION
**Pull request by Ralith**
This PR defines a stand-alone library with a simple API for persisting world state to disk, including both edited voxels and other entity states. It is motivated by these goals:

- Fast random reads to any spatial neighborhood, to allow streaming large saves (e.g. highly-edited worlds) on-demand for consistent performance
- Fast random writes of the same, allowing only modified parts of the world to be written when saving, minimizing gameplay disruption regardless of total save size
- Consistency even in the face of crashes or power loss
- Compactness, to support large worlds and minimize required disk I/O
- Forwards-compatibility to minimize the odds of old saves becoming unusable after an update
- Loose coupling with the rest of the game

The design space is large, and this proposal therefore makes some interesting choices:

### [redb](https://github.com/cberner/redb) as storage backend
Key-value stores provide fast random access to binary keys, which we can derive for nodes per https://github.com/Ralith/hypermine/issues/253. They also tend to offer strong durability guarantees. Several other implementations were considered:
- lmdb is highly mature, but has a legacy of second-class Windows support and historically poor Rust bindings. A good fallback option.
- RocksDB has strong industry backing, but poor Rust bindings, requiring libclang at build-time and requiring minutes(!) to build the underlying C++ library. It also may be more complex than our requirements merit.
- LevelDB is less actively developed than RocksDB, and appears to have no maintained Rust bindings.
- sled is pure-rust, but is immature, and has a history of data loss issues, cavalier attitudes to memory safety, and maintainer hostility. Maintenance also appears to be dropping off.

redb is immature, but is native Rust, actively developed, claims competitive performance, and has a simple API. Migration to another key-value store should be straightforward if necessary, but if redb proves up to the task it's an ideal solution--and we can afford to weather some issues and work with upstream to address them while hypermine itself is in early days.

Other alternatives considered:
- sqlite: extremely mature, would certainly never be a cause of data loss. However, it solves a lot of problems we don't have and hence is expected to have relatively poor performance.
- A custom key-value format with hand-rolled consistency mechanisms: Difficult to implement. In particular, supporting random writes while efficiently reusing space and guaranteeing consistency is complex. Higher risk of data loss bugs and likely inferior performance.

### Protocol Buffers for value encoding

Protocol Buffers are a mature, reasonably compact binary encoding with strong support for schema evolution. Fields can be added freely without compromising the capability to decode older values, or even the capability of older games to decode newer values. The prost crate provides an ergonomic interface for Rust.

These benefits come at the cost of slightly more space use than a minimal binary encoding, and requiring the foreign protoc compiler to be present in the build environment to process the schema. That requirement could be relaxed in the future by switching to manual, rather than build script driven, compilation of the schema.

Encoding schemes with similar features include Cap'n Proto, Apache Thrift, and Flatbuffers. Any would be basically adequate, but protobufs are the best supported.

### Opaque entity component storage

The protobuf schema makes no attempt to explicitly model component data like positions, instead storing them as opaque columns of binary data, one per type of component, each associated with a type tag. This allows these potentially small and large-volume values to be stored without per-field protobuf encoding overhead, while preserving forwards compatibility. If the representation of a component is changed, it will therefore need to be allocated a new ID.

### Flat voxel storage

A chunk's voxels are either not represented, or represented in full by a dense u16 array. Compared to more complex mechanisms like an explicit diff, this is trivial for the game to provide (by merely marking chunks as dirty when edited, and saving all dirty chunks). The enclosing layer of compression is expected to make this more than adequately compact, but the use of protobufs leaves the door open to backwards-compatibly introducing other representations in the future if desired.

### [zstd](https://github.com/facebook/zstd) compression

It's just the best general-purpose compression algorithm by a huge margin. Initial profiling shows time spent doing compression for writes is less than a quarter of the time spent inside redb. The current implementation uses the built-in zstd frame format, which adds ~2 bytes of overhead per node; this could be reduced to 1 byte with some additional effort using the raw zstd block API.

### CODE REVIEW

**Comment by Ralith** [comment]
I think this still needs to separately record a path from the origin to each player so that player positions can be restored when loading a save. We can't rely in node IDs for player locations because they're insufficient to relate nodes to eachother; we need a complete graph path as a seed on which neighborhood queries can build.

**Comment by patowen** [comment]
- [ ] We'll need to remember to include `protoc` in the build instructions.

**Comment by Ralith** [comment]
Reworked to use pre-generated protos so the build doesn't get more complicated. CI will verify that the protos are up to date.

**Comment by Ralith** [comment]
Now compressing the `Meta` proto as well.

File: save/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        let nodes = tx.open_table(NODE_TABLE)?;
+        let Some(node) = nodes.get(&node_id)? else { return Ok(None); };
+        Ok(Some(Node::decode(node.value())?))
+    }
+
+    pub fn put(&self, node_id: u128, node_state: Node) -> Result<(), redb::Error> {
+        let msg = node_state.encode_to_vec();
+        let tx = self.db.begin_write()?;
+        let mut nodes = tx.open_table(NODE_TABLE)?;
+        nodes.insert(node_id, &*msg)?;
```
Comment (patowen) [inline save/src/lib.rs]: For my own curiosity, what are the conventions behind `&*` vs `as_ref()`?

File: save/src/lib.rs [hunk 2]
```diff
@@ -174,7 +183,24 @@ impl Writer<'_, '_> {
...
+        self.meta_value.encode(&mut self.plain).unwrap();
+        self.compressed.clear();
+        self.compressed
+            .reserve(zstd::compress_bound(self.plain.len()));
+        self.cctx
+            .compress2(&mut self.compressed, &self.plain)
+            .map_err(zstd::get_error_name)
+            .unwrap();
+
+        self.meta.insert(&[][..], &*self.compressed)?;
```
Comment (patowen) [inline save/src/lib.rs:203]: This code looks the same as the code for `put` along with some of the code in `Save::open`. Should we extract this into a function?

**Review by patowen — CHANGES_REQUESTED** [review]
On my initial pass, this definitely seems like a good implementation of saving. Thanks for putting this all together! I looked through all the code, but please let me know if there are any parts you would like me to review with additional scrutiny.

File: save/src/lib.rs [hunk 3]
```diff
@@ -46,18 +46,66 @@ impl Save {
...
+
+impl ReaderGuard<'_> {
+    pub fn get(&self) -> Result<Reader<'_>, redb::Error> {
+        let table = self.tx.open_table(NODE_TABLE)?;
+        Ok(Reader { table })
+    }
+}
+
+pub struct Reader<'a> {
+    table: redb::ReadOnlyTable<'a, u128, &'static [u8]>,
```
Comment (patowen) [inline save/src/lib.rs:97]: I'm surprised the lifetime is `'static` here. It seems suspicious, as I'm not sure how ReDb can possibly expose a `'static` array reference without keeping it in memory forever.

File: save/src/protos.proto [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    // Number of voxels along the edge of a chunk
+    uint32 chunk_size = 1;
+}
+
+message Node {
+    // Entities whose origins lie within this node
+    repeated Archetype archetypes = 1;
+
+    // Voxel data for each modified chunk
+    repeated Chunk chunks = 2;
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/protos.proto:27]: All fields in a proto3 schema can be omitted (see https://protobuf.dev/programming-guides/proto3/#specifying-field-rules). I do generally intend to work in node space by default for stuff that isn't voxel data, regardless.

File: save/tests/tests.rs [hunk 5]
```diff
@@ -9,3 +9,27 @@ fn persist_meta() {
...
 }
+
+#[test]
+fn persist_node() {
+    let file = tempfile::NamedTempFile::new().unwrap();
+    let mut save = Save::open(file.path(), 12).unwrap();
+    let node = save::Node {
+        archetypes: vec![save::Archetype {
+            entities: vec![1, 2, 3],
+            component_types: vec![4, 5, 6],
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/tests/tests.rs:20]: It's 100% intended that unknown component types can be represented and manipulated, so I think this is a reasonable test case.

File: save/src/protos.proto [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    // Type of components stored in each column
+    repeated ComponentType component_types = 2;
+
+    // Columns of component data
+    repeated bytes component_data = 3;
+}
+
+message Chunk {
+    // Which dodecahedron vertex is associated with this chunk
+    uint32 vertex = 1;
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/protos.proto:44]: Protos don't have smaller integer types (see https://protobuf.dev/programming-guides/proto3/#scalar); even enums are just a thin cosmetic wrapper over `int32`. Instead, the variable-length "varint" encoding is used for all non-`fixed` integers, which will in practice always represent such small values with a single byte.

A previous draft defined vertex/side enums mirroring those of dodeca.rs, but I dropped those for two reasons:
- `prost` still ultimately works in integer terms so that unknown values can be represented, since enums in protos can generally expand in the future.
- `protoc` doesn't allow enums in the same scope to reuse the same name, so we couldn't have both `Vertex::A` and `Side::A`.

File: save/src/protos.proto [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+// A set of entities, all of which have the same components
+message Archetype {
+    // Entity IDs
+    repeated fixed64 entities = 1;
+
+    // Type of components stored in each column
+    repeated ComponentType component_types = 2;
+
+    // Columns of component data
+    repeated bytes component_data = 3;
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/protos.proto:39]: > Should it have the same number of elements as component_types?

Yep. A separate `Column` message would make this a bit clearer, but add extra indirection to the binary encoding.

> are a given component of each entry put together and compressed into a byte array?

Not separately compressed; that'd be redundant, since the whole `Node` gets compressed anyway. Just densely packed according to some component-specific serialization rule (e.g. dense `f32`s if `POSITION` ends up corresponding to a matrix).

File: save/benches/bench.rs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    let mut write = c.benchmark_group("write");
+    let node = save::Node {
+        archetypes: vec![save::Archetype {
+            entities: vec![1, 2, 3],
+            component_types: vec![4, 5, 6],
+            component_data: vec![],
+        }],
+        chunks: vec![save::Chunk {
+            vertex: 0,
+            voxels: vec![0; 12 * 12 * 2],
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/benches/bench.rs:18]: This isn't intended to be 100% representative, but rather just get us a ballpark frame of reference to start with and allow tweaks to be tested. In practice most chunk data should tend to have a *lot* of duplication, so this shouldn't be too far off anyway.

File: save/src/lib.rs [hunk 9]
```diff
@@ -14,13 +14,29 @@ pub struct Save {
...
+                }
+                Err(redb::Error::TableDoesNotExist(_)) => {
+                    let defaults = Meta {
+                        chunk_size: default_chunk_size.into(),
+                    };
+                    let tx = db.begin_write()?;
+                    let mut meta = tx.open_table(META_TABLE)?;
+                    meta.insert(&[][..], &*defaults.encode_to_vec())?;
+                    drop(meta);
+                    tx.open_table(NODE_TABLE)?; // Ensure reads don't fail oddly on empty saves
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/lib.rs:45]: Yep, exactly.

File: save/src/lib.rs (hunk 1 above)
Comment (Ralith) [reply save/src/lib.rs]: No strong consensus AFAIK.

File: save/src/lib.rs (hunk 3 above)
Comment (Ralith) [reply save/src/lib.rs:96]: Good instincts. In this case the `RedbValue` trait has a GAT that swaps out `'static` with the appropriate lifetime when accessing a value, so `'static` is really just a placeholder. The `&[u8]` used in e.g. the `const` table definitions is inferred to be the same.

File: save/src/lib.rs (hunk 2 above)
Comment (Ralith) [reply save/src/lib.rs:188]: Done.

File: save/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/lib.rs]: Reworded comment for clarity.

File: save/src/protos.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/protos.proto:40]: Reworded comment for clarity.

File: save/tests/tests.rs [hunk 10]
```diff
@@ -9,3 +9,27 @@ fn persist_meta() {
...
+            component_data: vec![],
+        }],
+        chunks: vec![save::Chunk {
+            vertex: 0,
+            voxels: vec![0; 12 * 12 * 2],
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/tests/tests.rs]: Fixed.

File: save/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "save"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
```
[the remark this answers is not in the document]
Comment (Ralith) [reply save/Cargo.toml:8]: Fixed.

**Comment by Ralith** [comment]
> please let me know if there are any parts you would like me to review with additional scrutiny.

The most important stuff is the design and tradeoffs (as discussed in the PR description) rather than the fairly mundane specifics of the code, though a careful eye on schema design is always justified. Of course, it's also important that the interface seems straightforwardly usable to you.

One additional noteworthy design decision is the choice to use a flat list of players, rather than introducing a new key-value table. While this should be fine for hundreds or even thousands of concurrent players, if we preserve player state indefinitely then it might grow without bound on a long-lived server. Some sort of pruning logic might be necessary. We could buy additional time by storing players as rows in a separate key-value table, so the whole set doesn't have to be loaded at once, but even then we'd still want pruning, and with pruning I imagine it'll necessarily remain small. Not sure what such a pruning scheme would look like, though... but this shouldn't be hard to change in the future.

File: save/src/protos.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (patowen) [reply save/src/protos.proto:40]: Do we want to edit the test case so that `component_types` and `component_data` have the same number of elements? Based on the reworded comment, it looks like the test case has an archetype with invalid data.

File: save/tests/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (patowen) [reply save/tests/tests.rs:20]: Fair point. It's essentially a question of whether we want to also test the standard path when no edge cases are in play.

**Comment by patowen** [comment]
> One additional noteworthy design decision is the choice to use a flat list of players, rather than introducing a new key-value table. While this should be fine for hundreds or even thousands of concurrent players, if we preserve player state indefinitely then it might grow without bound on a long-lived server. Some sort of pruning logic might be necessary. We could buy additional time by storing players as rows in a separate key-value table, so the whole set doesn't have to be loaded at once, but even then we'd still want pruning, and with pruning I imagine it'll necessarily remain small. Not sure what such a pruning scheme would look like, though... but this shouldn't be hard to change in the future.

That's a good call-out. I'm definitely a bit concerned that this could cause us trouble. How often do we want to update characters' positions in the save file? This could easily become the bottleneck, as any player far away from the origin would have a node path with a size measured in kilobytes. However, I think I had the wrong impression about save files in general. There's no reason we would want to save every second or every time someone places a block. We probably instead want to keep things in memory followed by saving all the changes in a single transaction once a minute or so.

With that in mind, if a player spends a whole day traveling at walking speed, which is about one node every 5 seconds, their node path would have a length of about 17000, which, compressed, should be about 8 KiB of data (assuming the compression can tell that all numbers are between 0 and 15). With 100 players doing that, it would be 800 KiB per minute of saving metadata, which is less than 1 MiB per minute. Computers can deal with way more than 1 MiB per _second_, so with this back-of-the-envelope calculation, this seems okay after all. I'm assuming that saving won't block the server from continuing to run.

The design as a whole definitely seems like it should work. There's definitely a chance that when I look into implementing saving, I realize that something important is missing, but it shouldn't be too difficult to fix then. While it does seem slightly tricky that we don't get full type-safety guarantees for invariants we want to hold (like character paths leading to the node with the appropriate archetype or `component_types` and `component_data` being the same length), but I don't consider this to be a major disadvantage.

File: save/src/protos.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/protos.proto:40]: These structures have no semantic import as far as this crate is concerned, so I don't think this is necessary; it will not impact the behavior the test is exercising. Unless it's sufficiently confusing as to compromise readability?

[the remark this answers is not in the document]
Comment (patowen) [reply save/src/protos.proto:40]: Right. I think readability is the main potential concern. When I left my latest comment, I believed there was a small chance that I misunderstood, and there were actually cases where the two lists should have a different number of elements. Since unit tests can often serve as a form of documentation, I was treating it as such.

However, I also agree that complicating the test data in a way that doesn't actually affect execution in any way is not ideal, so leaving a comment in the test noting that "`component_data` is left empty artificially for test simplicity despite it being against the spec" would likely avoid this confusion.

**Comment by Ralith** [comment]
> However, I think I had the wrong impression about save files in general. There's no reason we would want to save every second or every time someone places a block. We probably instead want to keep things in memory followed by saving all the changes in a single transaction once a minute or so.

Right. We actually might be able to get away with continuous saves (writing a single node takes <2ms) but batching things up is far more efficient.

> their node path would have a length of about 17000, which, compressed, should be about 8 KiB of data (assuming the compression can tell that all numbers are between 0 and 15)

zstd is very smart, so it should. A quick test using 17000 randomly generated edges in [0, 15) indeed yields 8.35KiB of data. Asymmetric numeral systems are good stuff.

> I'm assuming that saving won't block the server from continuing to run.

Good thought; I hadn't considered that. We can definitely generate a big pile of proto values and pass those off to a background thread to write them, which means we have *way* more margin for performance than I was planning for. Of course, life is a bit easier if we don't have to.

That said, your back-of-the-envelope example there indicates I drastically underestimated how big paths were liable to be. If individual paths are liable to be in the KiB range, the proportional overhead of a new row in a redb table is much less, and the memory and I/O cost of rewriting the entire set of players on each update is much higher. I'm therefore inclined to refactor for individual rows. With async saving it probably doesn't matter too much, but it'll make me happier at least.

> While it does seem slightly tricky that we don't get full type-safety guarantees for invariants we want to hold

This is a fact of life as soon as your data structures get nontrivial, unfortunately. Even in dependently typed languages where you could theoretically prove every invariant, it's not usually a good use of time.

> component_types and component_data being the same length

We could actually enforce this pretty easily by adding a `message ArchetypeColumn { uint32 type = 1; bytes data = 2; }` and having a list of those, but the extra indirection that results in the format bothers me cosmetically. Rationally it won't matter, but this particular straightforward issue doesn't seem likely to bite us. No matter what we do there, you could still e.g. have garbage data in the `data`.

**Comment by patowen** [comment]
> This is a fact of life as soon as your data structures get nontrivial, unfortunately. Even in dependently typed languages where you could theoretically prove every invariant, it's not usually a good use of time.

> We could actually enforce this pretty easily by adding a `message ArchetypeColumn { uint32 type = 1; bytes data = 2; }` and having a list of those, but the extra indirection that results in the format bothers me cosmetically. Rationally it won't matter, but this particular straightforward issue doesn't seem likely to bite us. No matter what we do there, you could still e.g. have garbage data in the `data`.

I'm definitely in agreement here. I haven't used protocol buffers at all, so I don't have a preferred style on how to organize them, but given the fundamental truth that we cannot achieve some arbitrary sense of "perfection", this design feels right to me.

After I give the code one more scan, I'm ready to approve this PR regardless of your decision regarding https://github.com/Ralith/hypermine/pull/266#discussion_r1130269444.

**Comment by patowen** [comment]
> I'm therefore inclined to refactor for individual rows. With async saving it probably doesn't matter too much, but it'll make me happier at least.

I can get behind that decision. Having all player positions get rewritten every time, even when they're not logged in, could potentially cause a performance drop for a highly counterintuitive reason, since I don't think most people would expect offline players to impact performance in any way.

**Review by patowen — APPROVED** [review]

File: save/tests/tests.rs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+#[test]
+fn persist_character() {
+    let file = tempfile::NamedTempFile::new().unwrap();
+    let mut save = Save::open(file.path(), 12).unwrap();
+    let mut writer_guard = save.write().unwrap();
+    let mut writer = writer_guard.get().unwrap();
+    let mut rng = SmallRng::from_entropy();
+    let mut path = Vec::with_capacity(17000);
+    for _ in 0..17000 {
+        path.push(rng.gen_range(0..15));
```
Comment (patowen) [inline save/tests/tests.rs:54]: I believe this should be `0..12`. I said "between 0 and 15" in my earlier comment because I was thinking of numbers that could be stored in half a byte.

File: save/src/protos.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ralith) [reply save/src/protos.proto:33]: Easier to fill it in.

File: save/tests/tests.rs (hunk 12 above)
Comment (Ralith) [reply save/tests/tests.rs]: Down to 7.9 KiB :tada:

Comment (Ralith) [reply save/tests/tests.rs]: Note zstd is very different from zlib, btw.

File: save/src/lib.rs (hunk 3 above)
Comment (patowen) [reply save/src/lib.rs]: Interesting. Thanks for the extra context. This definitely still confuses me a bit, but I should be able to learn this better once I find the time to look into it further.

File: save/tests/tests.rs (hunk 12 above)
Comment (patowen) [reply save/tests/tests.rs]: Oh, whoops. Thanks for the correction!
